mycpp/gc_heap.h

```diff
@@ -1,285 +1,37 @@
-    num_heap_growths_ = 0;
-    num_forced_growths_ = 0;
-    num_live_objs_ = 0;
-#endif
-  }
-
-  void* Bump(int n) {
-    char* p = free_;
-    free_ += n;
-#if GC_STATS
-    num_live_objs_++;
-#endif
-    return p;
-  }
-
-  void* Allocate(int num_bytes) {
-    int n = aligned(num_bytes);
-    // log("n = %d, p = %p", n, p);
-
-    // TODO(Jesse): Change to `assert(n >= sizeof(LayoutForwarded))`
-    //
-    // This must be at least sizeof(LayoutForwarded), which happens to be 16
-    // bytes, because the GC pointer forwarding requires 16 bytes.  If we
-    // allocated less than 16 the GC would overwrite the adjacent object when
-    // it went to forward the pointer.
-    assert(n >= 16);
-
-#if GC_EVERY_ALLOC
-    Collect();  // force collection to find problems early
-#endif
-
-    if (free_ + n <= limit_) {  // Common case: we have space for it.
-      return Bump(n);
-    }
-
-#if GC_STATS
-      // log("GC free_ %p,  from_space_ %p, space_size_ %d", free_, from_space_,
-      //    space_size_);
-#endif
-
-    Collect();  // Try to free some space.
-
-    // log("after GC: from begin %p, free_ %p,  n %d, limit_ %p",
-    //    from_space_.begin_, free_, n, limit_);
-
-    if (free_ + n <= limit_) {  // Now we have space for it.
-      return Bump(n);
-    }
-
-    // It's still too small.  Grow the heap.
-    int multiple = 2;
-    Collect((from_space_.size_ + n) * multiple);
-
-#if GC_STATS
-    num_forced_growths_++;
-#endif
-
-    return Bump(n);
-  }
-
-  void Swap() {
-    // Swap spaces for next collection.
-    char* tmp = from_space_.begin_;
-    from_space_.begin_ = to_space_.begin_;
-    to_space_.begin_ = tmp;
-
-    int tmp2 = from_space_.size_;
-    from_space_.size_ = to_space_.size_;
-    to_space_.size_ = tmp2;
-  }
-
-  void PushRoot(Obj** p) {
-    // log("PushRoot %d", roots_top_);
-    roots_[roots_top_++] = p;
-    // TODO: This should be like a malloc() failure?
-    assert(roots_top_ < kMaxRoots);
-  }
-
-  void PopRoot() {
-    roots_top_--;
-    // log("PopRoot %d", roots_top_);
-  }
-
-  Obj* Relocate(Obj* obj, Obj* header);
-
-  // mutates free_ and other variables
-  void Collect(int to_space_size = 0);
-
-#if GC_STATS
-  void Report() {
-    log("-----");
-    log("num collections = %d", num_collections_);
-    log("num heap growths = %d", num_heap_growths_);
-    log("num forced heap growths = %d", num_forced_growths_);
-    log("num live objects = %d", num_live_objs_);
-
-    log("from_space_ %p", from_space_.begin_);
-    log("to_space %p", to_space_.begin_);
-    log("-----");
-  }
-#endif
-
-  Space from_space_;  // space we allocate from
-  Space to_space_;    // space that the collector copies to
-
-  char* free_;   // next place to allocate, from_space_ <= free_ < limit_
-  char* limit_;  // end of space we're allocating from
+#define Terabytes(bytes) (Gigabytes(bytes) * 1024)
+#define Gigabytes(bytes) (Megabytes(bytes) * 1024)
+#define Megabytes(bytes) (Kilobytes(bytes) * 1024)
```

Comment:
This  is gonna sound real pedantic, but we should distinguish between powers of 2 (KiB, MiB, GiB) and powers of 10 (KB, MB, GB)

So I would change this to `MiB` and `KiB`   (and delete G and T if we're not using them)

Kilobytes are powers of 10, just like Kilograms are (yes this changed some time in the last 20 years)

Some people don't care about the difference, but when you're measuring storage and networking speeds, it leads to real bugs and real outages, so I always keep a strict distinction

(Also it is annoying when you're buying a hard drive, etc.)

Reply:
Changed.  I left `GiB` in because I'll certainly use it for testing in the future.